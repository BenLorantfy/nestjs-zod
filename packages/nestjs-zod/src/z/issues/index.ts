import {
  addIssueToContext,
  IssueData,
  ParseContext,
  ZodIssueOptionalMessage,
} from 'zod'
import { ZodAnyDateStringIssue } from './date-string'
import { ZodTooBigIssue, ZodTooSmallIssue } from './overrided'
import { ZodAnyPasswordIssue } from './password'

// eslint-disable-next-line @typescript-eslint/no-explicit-any
declare type StripPath<T extends object> = T extends any
  ? Omit<T, 'path'>
  : never

export type NestJsZodIssue = ZodAnyDateStringIssue | ZodAnyPasswordIssue

type ZodIssueOptionalMessageExtended =
  | ZodIssueOptionalMessage
  | NestJsZodIssue
  | ZodTooSmallIssue
  | ZodTooBigIssue

type ZodIssueExtended = ZodIssueOptionalMessageExtended & { message: string }

// for some reason "type" field breaks when using default Omit
type IssueDataExtended = StripPath<ZodIssueOptionalMessageExtended> & {
  path?: (string | number)[]
  fatal?: boolean
}

function addIssueToContextExtended(
  context: ParseContext,
  issueData: IssueDataExtended
) {
  addIssueToContext(context, issueData as IssueData)
}

export { addIssueToContextExtended as addIssueToContext }
export type { ZodIssueOptionalMessageExtended as ZodIssueOptionalMessage }
export type { ZodIssueExtended as ZodIssue }

export * from './date-string'
export * from './overrided'
export * from './password'
