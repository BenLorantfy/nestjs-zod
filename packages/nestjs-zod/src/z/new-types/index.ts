export * from './date-string'
export * from './password'
