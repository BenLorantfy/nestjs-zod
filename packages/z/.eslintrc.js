const { configure, presets } = require('eslint-kit')

module.exports = configure({
  presets: [
    presets.imports(),
    presets.node(),
    presets.prettier(),
    presets.typescript(),
  ],
  extend: {
    rules: {
      '@typescript-eslint/no-unnecessary-condition': 'off',
    },
  },
})
