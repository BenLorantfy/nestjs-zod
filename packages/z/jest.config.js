module.exports = {
  preset: 'ts-jest',
  testEnvironment: 'node',
  passWithNoTests: true
};