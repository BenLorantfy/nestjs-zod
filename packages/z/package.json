{
  "name": "@nest-zod/z",
  "description": "Extended zod with a few additional types and helpers",
  "version": "1.0.1",
  "main": "./dist/z.js",
  "exports": {
    ".": {
      "default": "./dist/z.js"
    },
    "./frontend": {
      "import": "./dist/frontend.mjs",
      "default": "./dist/frontend.js"
    }
  },
  "module": "./dist/z.js",
  "types": "./dist/z.d.ts",
  "files": [
    "dist",
    "z.d.ts",
    "frontend.d.ts"
  ],
  "sideEffects": false,
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git@github.com:BenLorantfy/nestjs-zod.git",
    "directory": "packages/z"
  },
  "author": "Evgeny Zakharov <to@risen.dev>",
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "dev": "vite",
    "build": "rollup -c",
    "test": "TZ=UTC jest",
    "lint": "eslint --ext .ts,.tsx src",
    "lint:fix": "eslint --ext .ts,.tsx src --fix"
  },
  "devDependencies": {
    "@golevelup/ts-jest": "^0.3.3",
    "@nestjs/common": "^10.0.0",
    "@nestjs/core": "^10.0.0",
    "@nestjs/swagger": "^7.4.2",
    "@types/jest": "^29.5.2",
    "@types/merge-deep": "^3.0.0",
    "@types/node": "^22.7.5",
    "esbuild": "^0.14.24",
    "eslint": "^8.42.0",
    "eslint-kit": "^5.7.0",
    "jest": "^29.5.0",
    "reflect-metadata": "^0.2.0",
    "rollup": "^2.69.0",
    "rollup-plugin-bundle-size": "^1.0.3",
    "rollup-plugin-copy": "^3.4.0",
    "rollup-plugin-dts": "^4.1.0",
    "rollup-plugin-esbuild": "^4.8.2",
    "rollup-plugin-terser": "^7.0.2",
    "rxjs": "^7.8.1",
    "ts-jest": "^29.1.0",
    "typescript": "^5.1.3",
    "zod": "^3.23.8"
  },
  "peerDependencies": {
    "@nestjs/common": "^10.0.0",
    "@nestjs/core": "^10.0.0",
    "@nestjs/swagger": "^7.4.2",
    "zod": ">= 3.14.3"
  },
  "peerDependenciesMeta": {
    "@nestjs/common": {
      "optional": true
    },
    "@nestjs/core": {
      "optional": true
    },
    "@nestjs/swagger": {
      "optional": true
    },
    "zod": {
      "optional": false
    }
  }
}