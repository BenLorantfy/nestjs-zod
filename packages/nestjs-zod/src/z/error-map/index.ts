import {
  defaultErrorMap,
  setErrorMap,
  ZodErrorMap,
  ZodIssueOptionalMessage as ZodIssueOptionalMessageDefault,
} from 'zod'
import { ZodIssueOptionalMessage } from '../issues'
import { dateStringCustom, dateStringYearMinMax } from './date-string'
import { passwordCustom, passwordMinMax } from './password'
import { composeMappers } from './shared'

type ErrorMapContext = Parameters<ZodErrorMap>[1]

type ZodErrorMapExtended = (
  issue: ZodIssueOptionalMessage,
  context: ErrorMapContext
) => ReturnType<ZodErrorMap>

const mapper = composeMappers([
  dateStringCustom,
  dateStringYearMinMax,
  passwordCustom,
  passwordMinMax,
])

const extendedErrorMap: ZodErrorMapExtended = (issue, context) => {
  const result = mapper(issue)

  if (result.matched) {
    return { message: result.message }
  }

  /*
   * At this moment, there are no custom Issues remained,
   * so we can safely delegate the control to defaultErrorMap
   */

  return defaultErrorMap(issue as ZodIssueOptionalMessageDefault, context)
}

function setExtendedErrorMap(map: ZodErrorMapExtended) {
  setErrorMap(map)
}

// set custom errorMap before any user code is executed
setExtendedErrorMap(extendedErrorMap)

export {
  extendedErrorMap as defaultErrorMap,
  setExtendedErrorMap as setErrorMap,
}

export type { ZodErrorMapExtended as ZodErrorMap }
