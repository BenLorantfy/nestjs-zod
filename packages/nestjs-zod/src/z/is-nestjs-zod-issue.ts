import { NestJsZodIssue, ZodIssueOptionalMessage } from './issues'

// Inline code to prevent bundling zod in frontend package
const CUSTOM_ISSUE_CODE = 'custom'

export function isNestJsZodIssue(
  issue: ZodIssueOptionalMessage
): issue is NestJsZodIssue {
  return issue.code === CUSTOM_ISSUE_CODE && issue.params?.isNestJsZod
}
