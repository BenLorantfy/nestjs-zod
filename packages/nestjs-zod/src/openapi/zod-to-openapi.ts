import { Type } from '@nestjs/common'
import { SchemaObject } from '@nestjs/swagger/dist/interfaces/open-api-spec.interface'
import mergeDeep from 'merge-deep'
import { z } from '../z'

interface ExtendedSchemaObject extends SchemaObject {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [key: `x-${string}`]: any
}

export function is<T extends Type<z.ZodTypeAny>>(
  input: z.ZodTypeAny,
  factory: T
): input is InstanceType<T> {
  const factories = z as unknown as Record<string, Type<z.ZodTypeAny>>
  return factory === factories[input._def.typeName]
}

export function zodToOpenAPI(
  zodType: z.ZodTypeAny,
  visited: Set<any> = new Set()
) {
  const object: ExtendedSchemaObject = {}

  if (zodType.description) {
    object.description = zodType.description
  }

  if (is(zodType, z.ZodString)) {
    const { checks } = zodType._def
    object.type = 'string'

    for (const check of checks) {
      if (check.kind === 'min') {
        object.minLength = check.value
      } else if (check.kind === 'max') {
        object.maxLength = check.value
      } else if (check.kind === 'email') {
        object.format = 'email'
      } else if (check.kind === 'url') {
        object.format = 'uri'
      } else if (check.kind === 'uuid') {
        object.format = 'uuid'
      } else if (check.kind === 'cuid') {
        object.format = 'cuid'
      } else if (check.kind === 'regex') {
        object.pattern = check.regex.source
      } else if (check.kind === 'datetime') {
        object.format = 'date-time'
      }
    }
  }

  if (is(zodType, z.ZodPassword)) {
    const { checks } = zodType._def
    const regex = zodType.buildFullRegExp()
    object.type = 'string'
    object.format = 'password'
    object.pattern = regex.source

    for (const check of checks) {
      if (check.kind === 'minLength') {
        object.minLength = check.value
      } else if (check.kind === 'maxLength') {
        object.maxLength = check.value
      }
    }
  }

  if (is(zodType, z.ZodBoolean)) {
    object.type = 'boolean'
  }

  if (is(zodType, z.ZodNumber)) {
    const { checks } = zodType._def
    object.type = 'number'

    for (const check of checks) {
      if (check.kind === 'int') {
        object.type = 'integer'
      } else if (check.kind === 'min') {
        object.minimum = check.value
        object.exclusiveMinimum = !check.inclusive
      } else if (check.kind === 'max') {
        object.maximum = check.value
        object.exclusiveMaximum = !check.inclusive
      } else if (check.kind === 'multipleOf') {
        object.multipleOf = check.value
      }
    }
  }

  if (is(zodType, z.ZodDateString)) {
    const { checks } = zodType._def
    object.type = 'string'

    for (const check of checks) {
      if (check.kind === 'format') {
        object.format = check.value
      }
    }
  }

  if (is(zodType, z.ZodBigInt)) {
    object.type = 'integer'
    object.format = 'int64'
  }

  if (is(zodType, z.ZodArray)) {
    const { minLength, maxLength, type } = zodType._def
    object.type = 'array'
    if (minLength) object.minItems = minLength.value
    if (maxLength) object.maxItems = maxLength.value
    object.items = zodToOpenAPI(type, visited)
  }

  if (is(zodType, z.ZodTuple)) {
    const { items } = zodType._def
    object.type = 'array'
    object.items = { oneOf: items.map((item) => zodToOpenAPI(item, visited)) }
  }

  if (is(zodType, z.ZodSet)) {
    const { valueType, minSize, maxSize } = zodType._def
    object.type = 'array'
    if (minSize) object.minItems = minSize.value
    if (maxSize) object.maxItems = maxSize.value
    object.items = zodToOpenAPI(valueType, visited)
    object.uniqueItems = true
  }

  if (is(zodType, z.ZodUnion)) {
    const { options } = zodType._def
    object.oneOf = options.map((option) => zodToOpenAPI(option, visited))
  }

  if (is(zodType, z.ZodDiscriminatedUnion)) {
    const { options } = zodType._def
    object.oneOf = []
    for (const schema of options.values()) {
      object.oneOf.push(zodToOpenAPI(schema, visited))
    }
  }

  if (is(zodType, z.ZodLiteral)) {
    const { value } = zodType._def

    if (typeof value === 'string') {
      object.type = 'string'
      object.enum = [value]
    }

    if (typeof value === 'number') {
      object.type = 'number'
      object.minimum = value
      object.maximum = value
    }

    if (typeof value === 'boolean') {
      // currently there is no way to completely describe boolean literal
      object.type = 'boolean'
    }
  }

  if (is(zodType, z.ZodEnum)) {
    const { values } = zodType._def
    object.type = 'string'
    object.enum = values
  }

  if (is(zodType, z.ZodNativeEnum)) {
    const { values } = zodType._def
    // this only supports enums with string literal values
    object.type = 'string'
    object.enum = Object.values(values)
    object['x-enumNames'] = Object.keys(values)
  }

  if (is(zodType, z.ZodTransformer)) {
    const { schema } = zodType._def
    Object.assign(object, zodToOpenAPI(schema, visited))
  }

  if (is(zodType, z.ZodNullable)) {
    const { innerType } = zodType._def
    Object.assign(object, zodToOpenAPI(innerType, visited))
    object.nullable = true
  }

  if (is(zodType, z.ZodOptional)) {
    const { innerType } = zodType._def
    Object.assign(object, zodToOpenAPI(innerType, visited))
  }

  if (is(zodType, z.ZodDefault)) {
    const { defaultValue, innerType } = zodType._def
    Object.assign(object, zodToOpenAPI(innerType, visited))
    object.default = defaultValue()
  }

  if (is(zodType, z.ZodObject)) {
    const { shape } = zodType._def
    object.type = 'object'

    object.properties = {}
    object.required = []

    for (const [key, schema] of Object.entries<z.ZodTypeAny>(shape())) {
      object.properties[key] = zodToOpenAPI(schema, visited)
      const optionalTypes = [z.ZodOptional.name, z.ZodDefault.name]
      const isOptional = optionalTypes.includes(schema.constructor.name)
      if (!isOptional) object.required.push(key)
    }

    if (object.required.length === 0) {
      delete object.required
    }
  }

  if (is(zodType, z.ZodRecord)) {
    const { valueType } = zodType._def
    object.type = 'object'
    object.additionalProperties = zodToOpenAPI(valueType, visited)
  }

  if (is(zodType, z.ZodIntersection)) {
    const { left, right } = zodType._def
    const merged = mergeDeep(
      zodToOpenAPI(left, visited),
      zodToOpenAPI(right, visited)
    )
    Object.assign(object, merged)
  }

  if (is(zodType, z.ZodEffects)) {
    const { schema } = zodType._def
    Object.assign(object, zodToOpenAPI(schema, visited))
  }

  if (is(zodType, z.ZodLazy)) {
    const { getter } = zodType._def
    if (visited.has(getter)) return object

    visited.add(getter)
    Object.assign(object, zodToOpenAPI(getter(), visited))
  }

  return object
}
