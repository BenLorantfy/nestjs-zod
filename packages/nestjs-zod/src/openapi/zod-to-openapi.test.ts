import { z, ZodTypeAny } from '../z'
import { zodToOpenAPI } from './zod-to-openapi'

const complexTestSchema = z.object({
  stringMinMax: z.string().min(5).max(15),
  numberMinMax: z.number().min(3).max(10),
  numberMultipleOf: z.number().multipleOf(10),
  enumDefault: z.enum(['buy-it-now', 'auctions']).default('buy-it-now'),
  tuple: z.tuple([
    z.string(),
    z.object({ name: z.literal('Rudy') }),
    z.array(z.union([z.literal('blue'), z.literal('red')])),
  ]),
  union: z
    .union([z.literal('kek'), z.literal('topkek')])
    .describe('Please choose topkek'),
  discriminatedUnion: z.discriminatedUnion('name', [
    z.object({ name: z.literal('vasya'), age: z.number() }),
    z.object({ name: z.literal('petya'), age: z.number() }),
  ]),
  literalString: z.literal('123'),
  literalNumber: z.literal(123),
  literalBoolean: z.literal(true),
  array: z.array(z.string()).max(3),
  objectNested: z.object({
    uuid: z.string().uuid(),
  }),
  record: z.record(z.number()),
  recordWithKeys: z.record(z.number(), z.string()),
  dateString: z.dateString().cast().describe('My date string'),
  zodDateString: z.string().datetime(),
  password: z.password(),
  passwordComplex: z
    .password()
    .atLeastOne('digit')
    .atLeastOne('uppercase')
    .min(8)
    .max(100),
})

const intersectedObjectsSchema = z.intersection(
  z.object({
    one: z.number(),
  }),
  z.object({
    two: z.number(),
  })
)

const intersectedUnionsSchema = z.intersection(
  z.union([z.literal('123'), z.number()]),
  z.union([z.boolean(), z.array(z.string())])
)

const overrideIntersectionSchema = z.intersection(
  z.object({
    one: z.number(),
  }),
  z.object({
    one: z.string(),
  })
)

const transformedSchema = z
  .object({
    seconds: z.number(),
  })
  .transform((value) => ({
    seconds: value.seconds,
    minutes: value.seconds / 60,
    hours: value.seconds / 3600,
  }))

const lazySchema = z.lazy(() => z.string())

it('should serialize a complex schema', () => {
  const openApiObject = zodToOpenAPI(complexTestSchema)

  expect(openApiObject).toMatchSnapshot()
})

it('should serialize an intersection of objects', () => {
  const openApiObject = zodToOpenAPI(intersectedObjectsSchema)

  expect(openApiObject).toMatchSnapshot()
})

it('should serialize an intersection of unions', () => {
  const openApiObject = zodToOpenAPI(intersectedUnionsSchema)

  expect(openApiObject).toMatchSnapshot()
})

it('should serialize an intersection with overrided fields', () => {
  const openApiObject = zodToOpenAPI(overrideIntersectionSchema)

  expect(openApiObject).toMatchSnapshot()
})

it('should serialize objects', () => {
  const schema = z.object({
    prop1: z.string(),
    prop2: z.string().optional(),
  })
  const openApiObject = zodToOpenAPI(schema)

  expect(openApiObject).toEqual({
    type: 'object',
    required: ['prop1'],
    properties: {
      prop1: {
        type: 'string',
      },
      prop2: {
        type: 'string',
      },
    },
  })
})

it('should serialize partial objects', () => {
  const schema = z
    .object({
      prop1: z.string(),
      prop2: z.string(),
    })
    .partial()
  const openApiObject = zodToOpenAPI(schema)

  expect(openApiObject).toEqual({
    type: 'object',
    properties: {
      prop1: {
        type: 'string',
      },
      prop2: {
        type: 'string',
      },
    },
  })
})

it('should serialize nullable types', () => {
  const schema = z.string().nullable()
  const openApiObject = zodToOpenAPI(schema)

  expect(openApiObject).toEqual({ type: 'string', nullable: true })
})

it('should serialize optional types', () => {
  const schema = z.string().optional()
  const openApiObject = zodToOpenAPI(schema)

  expect(openApiObject).toEqual({ type: 'string' })
})

it('should serialize types with default value', () => {
  const schema = z.string().default('abitia')
  const openApiObject = zodToOpenAPI(schema)

  expect(openApiObject).toEqual({ type: 'string', default: 'abitia' })
})

it('should serialize enums', () => {
  const schema = z.enum(['adama', 'kota'])
  const openApiObject = zodToOpenAPI(schema)

  expect(openApiObject).toEqual({
    type: 'string',
    enum: ['adama', 'kota'],
  })
})

it('should serialize native enums', () => {
  enum NativeEnum {
    ADAMA = 'adama',
    KOTA = 'kota',
  }

  const schema = z.nativeEnum(NativeEnum)
  const openApiObject = zodToOpenAPI(schema)

  expect(openApiObject).toEqual({
    'type': 'string',
    'enum': ['adama', 'kota'],
    'x-enumNames': ['ADAMA', 'KOTA'],
  })
})

describe('scalar types', () => {
  const testCases: [ZodTypeAny, string, string?][] = [
    // [zod type, expected open api type, expected format]
    [z.string(), 'string'],
    [z.number(), 'number'],
    [z.boolean(), 'boolean'],
    [z.bigint(), 'integer', 'int64'],
    // [z.null(), 'null'], <- Needs OpenApi 3.1 to be represented correctly
    // [z.undefined(), 'undefined'], <- TBD, probably the property should be removed from schema
  ]

  for (const [zodType, expectedType, expectedFormat] of testCases) {
    // eslint-disable-next-line no-loop-func
    it(expectedType, () => {
      const openApiObject = zodToOpenAPI(zodType)

      expect(openApiObject).toEqual({
        type: expectedType,
        format: expectedFormat ?? undefined,
      })
    })
  }
})

it('should serialize transformed schema', () => {
  const openApiObject = zodToOpenAPI(transformedSchema)

  expect(openApiObject).toEqual({
    type: 'object',
    required: ['seconds'],
    properties: {
      seconds: {
        type: 'number',
      },
    },
  })
})

it('should serialize lazy schema', () => {
  const openApiObject = zodToOpenAPI(lazySchema)

  expect(openApiObject).toEqual({
    type: 'string',
  })
})
