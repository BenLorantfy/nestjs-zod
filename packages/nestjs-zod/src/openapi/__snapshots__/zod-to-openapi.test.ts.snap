// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should serialize a complex schema 1`] = `
Object {
  "properties": Object {
    "array": Object {
      "items": Object {
        "type": "string",
      },
      "maxItems": 3,
      "type": "array",
    },
    "dateString": Object {
      "description": "My date string",
      "format": "date-time",
      "type": "string",
    },
    "discriminatedUnion": Object {
      "oneOf": Array [
        Object {
          "properties": Object {
            "age": Object {
              "type": "number",
            },
            "name": Object {
              "enum": Array [
                "vasya",
              ],
              "type": "string",
            },
          },
          "required": Array [
            "name",
            "age",
          ],
          "type": "object",
        },
        Object {
          "properties": Object {
            "age": Object {
              "type": "number",
            },
            "name": Object {
              "enum": Array [
                "petya",
              ],
              "type": "string",
            },
          },
          "required": Array [
            "name",
            "age",
          ],
          "type": "object",
        },
      ],
    },
    "enumDefault": Object {
      "default": "buy-it-now",
      "enum": Array [
        "buy-it-now",
        "auctions",
      ],
      "type": "string",
    },
    "literalBoolean": Object {
      "type": "boolean",
    },
    "literalNumber": Object {
      "maximum": 123,
      "minimum": 123,
      "type": "number",
    },
    "literalString": Object {
      "enum": Array [
        "123",
      ],
      "type": "string",
    },
    "numberMinMax": Object {
      "exclusiveMaximum": false,
      "exclusiveMinimum": false,
      "maximum": 10,
      "minimum": 3,
      "type": "number",
    },
    "numberMultipleOf": Object {
      "multipleOf": 10,
      "type": "number",
    },
    "objectNested": Object {
      "properties": Object {
        "uuid": Object {
          "format": "uuid",
          "type": "string",
        },
      },
      "required": Array [
        "uuid",
      ],
      "type": "object",
    },
    "password": Object {
      "format": "password",
      "pattern": "^.*$",
      "type": "string",
    },
    "passwordComplex": Object {
      "format": "password",
      "maxLength": 100,
      "minLength": 8,
      "pattern": "^(?:(?=.*\\\\d)(?=.*[A-Z]).*)$",
      "type": "string",
    },
    "record": Object {
      "additionalProperties": Object {
        "type": "number",
      },
      "type": "object",
    },
    "recordWithKeys": Object {
      "additionalProperties": Object {
        "type": "string",
      },
      "type": "object",
    },
    "stringMinMax": Object {
      "maxLength": 15,
      "minLength": 5,
      "type": "string",
    },
    "tuple": Object {
      "items": Object {
        "oneOf": Array [
          Object {
            "type": "string",
          },
          Object {
            "properties": Object {
              "name": Object {
                "enum": Array [
                  "Rudy",
                ],
                "type": "string",
              },
            },
            "required": Array [
              "name",
            ],
            "type": "object",
          },
          Object {
            "items": Object {
              "oneOf": Array [
                Object {
                  "enum": Array [
                    "blue",
                  ],
                  "type": "string",
                },
                Object {
                  "enum": Array [
                    "red",
                  ],
                  "type": "string",
                },
              ],
            },
            "type": "array",
          },
        ],
      },
      "type": "array",
    },
    "union": Object {
      "description": "Please choose topkek",
      "oneOf": Array [
        Object {
          "enum": Array [
            "kek",
          ],
          "type": "string",
        },
        Object {
          "enum": Array [
            "topkek",
          ],
          "type": "string",
        },
      ],
    },
    "zodDateString": Object {
      "format": "date-time",
      "type": "string",
    },
  },
  "required": Array [
    "stringMinMax",
    "numberMinMax",
    "numberMultipleOf",
    "tuple",
    "union",
    "discriminatedUnion",
    "literalString",
    "literalNumber",
    "literalBoolean",
    "array",
    "objectNested",
    "record",
    "recordWithKeys",
    "dateString",
    "zodDateString",
    "password",
    "passwordComplex",
  ],
  "type": "object",
}
`;

exports[`should serialize an intersection of objects 1`] = `
Object {
  "properties": Object {
    "one": Object {
      "type": "number",
    },
    "two": Object {
      "type": "number",
    },
  },
  "required": Array [
    "one",
    "two",
  ],
  "type": "object",
}
`;

exports[`should serialize an intersection of unions 1`] = `
Object {
  "oneOf": Array [
    Object {
      "enum": Array [
        "123",
      ],
      "type": "string",
    },
    Object {
      "type": "number",
    },
    Object {
      "type": "boolean",
    },
    Object {
      "items": Object {
        "type": "string",
      },
      "type": "array",
    },
  ],
}
`;

exports[`should serialize an intersection with overrided fields 1`] = `
Object {
  "properties": Object {
    "one": Object {
      "type": "string",
    },
  },
  "required": Array [
    "one",
  ],
  "type": "object",
}
`;
